nixos/modules/image/amend-repart-definitions.py

```diff
@@ -0,0 +1,113 @@
+#!/usr/bin/env python
+
+"""Amend systemd-repart definiton files.
+
+In order to avoid Import-From-Derivation (IFD) when building images with
+systemd-repart, the definition files created by Nix need to be amended with the
+store paths from the closure.
+
+This is achieved by adding CopyFiles= instructions to the definition files.
+
+The arbitrary files configured via `contents` are also added to the definition
+files using the same mechanism.
+"""
+
+import json
+import sys
+import shutil
+import os
+import tempfile
+from pathlib import Path
+
+
+def add_contents_to_definition(
+    definition: Path, contents: dict[str, dict[str, str]] | None
+) -> None:
+    """Add CopyFiles= instructions to a definition for all files in contents."""
+    if not contents:
+        return
+
+    copy_files_lines: list[str] = []
+    for target, options in contents.items():
+        source = options["source"]
+
+        copy_files_lines.append(f"CopyFiles={source}:{target}\n")
+
+    with open(definition, "a") as f:
+        f.writelines(copy_files_lines)
+
+
+def add_closure_to_definition(
+    definition: Path, closure: Path | None, strip_nix_store_prefix: bool | None
+) -> None:
+    """Add CopyFiles= instructions to a definition for all paths in the closure.
+
+    If strip_nix_store_prefix is True, `/nix/store` is stripped from the target path.
+    """
+    if not closure:
+        return
+
+    copy_files_lines: list[str] = []
+    with open(closure, "r") as f:
```

Comment:
```suggestion
    nix_path_registration_source = closure+"/registration"
    nix_path_registration_target = "nix-path-registration" if strip_nix_store_prefix else "/nix/store/nix-path-registration"
    copy_file_lines.append(f"CopyFiles={nix_path_registration_source}:{nix_path_registration_target}")
    with open(closure+"/store-paths", "r") as f:
```

Reply:
Your suggested solution has the downside that it puts the registration into the nix store. We don't actually have to touch the python script to achieve get the registration into our image though.

You can just do `contents."/registration".source = "${closure}/registration"`. Because the closure is a derivation, this would just use the cached derivation and not actually compute anything new. 

Now you can also pick where the registration ends up in your image. E.g. it could be on the root partition while the nix store (potentially dm-verity backed) is on a completely different partition.

On a side note: I'd really want to avoid to proliferate `boot.postBootCommands` in Nixpkgs because I'd love to see scripted activation be replaced by parallel services that perform the activation. Of course you could achieve the same if you just write a service yourself.

Reply:
Cool! Clear! I might make a follow-up PR to expose this as an option then (and try to come up with something different than `postBootCommands`).